lib/Runtime/Base/ThreadContext.cpp

```diff
@@ -3551,21 +3598,110 @@ ThreadContext::InvalidatePropertyGuardEntry(const Js::PropertyRecord* propertyRe
 
     entry->uniqueGuards.Clear();
 
-
     // Count no. of invalidations done so far. Exclude if this is all property guards invalidation in which case
     // the unique Guards will be cleared anyway.
     if (!isAllPropertyGuardsInvalidation)
     {
-        this->recyclableData->constructorCacheInvalidationCount += count;
-        if (this->recyclableData->constructorCacheInvalidationCount > (uint)CONFIG_FLAG(ConstructorCacheInvalidationThreshold))
+        this->UpdateConstructorCacheInvalidationCount(count);
+    }
+
+    this->UpdateEntryPoints(propertyRecord, entry);
+}
+
+void
+ThreadContext::InvalidatePropertyGuards(Js::PropertyId propertyId)
+{
+    const Js::PropertyRecord* propertyRecord = GetPropertyName(propertyId);
+    PropertyGuardDictionary &guards = this->recyclableData->propertyGuards;
+    PropertyGuardEntry* entry = nullptr;
+    if (guards.TryGetValueAndRemove(propertyRecord, &entry))
+    {
+        InvalidatePropertyGuardEntry(propertyRecord, entry, false);
+    }
+}
+
+void
+ThreadContext::InvalidatePropertyGuardEntryForType(const Js::PropertyRecord* propertyRecord, PropertyGuardEntry* entry, bool isAllPropertyGuardsInvalidation, const Js::Type *type)
+{
+    Assert(entry != nullptr);
+    Assert(type != nullptr);
+
+    if (entry->sharedGuard != nullptr)
+    {
+        Js::PropertyGuard* guard = entry->sharedGuard;
+        if (guard->GetValue() == reinterpret_cast<intptr_t>(type))
+        {
+            if (PHASE_TRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, address: 0x%p, value: 0x%p, value address: 0x%p\n"),
+                              propertyRecord->GetBuffer(), guard, guard->GetValue(), guard->GetAddressOfValue());
+                Output::Flush();
+            }
+
+            if (PHASE_TESTTRACE1(Js::TracePropertyGuardsPhase) || PHASE_VERBOSE_TESTTRACE1(Js::FixedMethodsPhase))
+            {
+                Output::Print(_u("FixedFields: invalidating guard: name: %s, value: 0x%p\n"), propertyRecord->GetBuffer(), guard->GetValue());
+                Output::Flush();
+            }
+
+            guard->Invalidate();
+        }
+    }
+
+    uint count = 0;
+    entry->uniqueGuards.MapAndRemoveIf([&count, propertyRecord, type](RecyclerWeakReference<Js::PropertyGuard>* guardWeakRef) -> bool
+    {
+        Js::PropertyGuard* guard = guardWeakRef->Get();
+        if (guard != nullptr && guard->GetValue() == reinterpret_cast<intptr_t>(type))
```

Comment:
Should I remove entries where guard == nullptr?  Again, this seems like unnecessary cleanup, but I'm open to discussion.

Reply:
s/unnecessary/unrelated/

Reply:
Seems reasonable. The existing method InvalidatePropertyGuardEntry completely clears the map (including null entries), so I guess there's some precedent for removing them.

---
In reply to: [172374425](https://github.com/Microsoft/ChakraCore/pull/4780#discussion_r172374425) [](ancestors = 172374425)